src/memo/memo.ml

```diff
@@ -704,16 +680,150 @@ module Call_stack = struct
+        | true ->
+          (* Here we know that the current [frame] has already been traversed in
+             a previous [add_path_to] call. Therefore, the DAG already contains
+             all the edges that we will discover by continuing the recursive
+             traversal. We might as well stop here and save time. *)
+          Ok ()
+        | false -> (
+          let caller_dag_node =
+            Lazy.force (Stack_frame_with_state.dag_node frame)
+          in
+          match Dag.add_assuming_missing caller_dag_node dag_node with
+          | exception Dag.Cycle cycle ->
+            Error (List.map cycle ~f:(fun dag_node -> dag_node.Dag.data))
+          | () -> (
+            if !Counters.enabled then
+              incr Counters.edges_in_cycle_detection_graph;
+            let not_traversed_before =
+              Dag.Id.Set.is_empty children_added_to_dag
+            in
+            Stack_frame_with_state.record_child_added_to_dag frame ~dag_node_id;
+            match not_traversed_before with
+            | true -> add_path_impl stack caller_dag_node
+            | false ->
+              (* Same optimisation as above: no need to traverse again. *)
+              Ok ())))
+    in
+    add_path_impl stack (Lazy.force dag_node)
+
+  (* Add a dependency on the [dep_node] from the caller, if there is one. *)
+  let add_dep_from_caller =
+    let get_call_stack_tip () = get_call_stack () >>| List.hd_opt in
+    fun dep_node ->
+      let+ caller = get_call_stack_tip () in
+      match caller with
+      | None -> ()
+      | Some caller -> Stack_frame_with_state.add_dep caller ~dep_node
+end
+
+(* This module contains the essence of our cycle detection algorithm. Briefly,
+   the idea is as follows: whenever we are about to get blocked to wait for the
+   result of a computation that is currently running, we add the current call
+   stack to the cycle detection DAG. If this creates a cycle, we stop and report
+   a "dependency cycle" error; otherwise, we proceed with the blocking.
+
+   Below are some notes on how/why this algorithm works.
+
+   By "computation", we mean execution of a fiber. To share computation results
+   between multiple concurrent readers, we use [Once.t], which is a thin wrapper
+   around [Fiber.t] that executes the underlying fiber at most once. The first
+   reader "forces" the fiber and eventually gets the result, while other readers
+   subscribe to the future result of the computation via a [Fiber.Ivar.t]. This
+   blocks their execution and can lead to a deadlock if there is a dependency
+   cycle between different computations. To check for cycles *before* getting
+   blocked, a reader can use [Once.force_with_blocking_check].
+
+   One simple algorithm to check for cycles is to create a DAG node for every
+   computation, and add a DAG edge whenever a computation would like to read the
+   result of another unfinished computation. If adding an edge creates a cycle,
```

Comment:
I'm not sure if the definitions you gave above take care of it (I think no because whether or not a "future execution of a fiber" is an "execution of a fiber" is a philosophical question that I have no answer for), but in the vacuum the phrase "unfinished computation" suggests to me that the computation was already started, even though the key point here is that we do it even for the ones that haven't started yet. Feels potentially confusing.

Also, there's another optimization involved here that we're just skipping over. Maybe it's worth a passing mention if we're starting from the naive approach.

I wanted to propose an alternative term to "computation", such as "computation node", or even "thunk",
but maybe this tweak is enough at the point where we introduce it:

```
By "computation", we mean a shared execution of a fiber, represented with [Once.t].
Such computations can be in one of three states: not started, running, and finished.
```

Then we could change the paragraph below like this:

```
One simple algorithm to check for cycles is to create a DAG node for every
computation, and add a DAG edge whenever that computation would like to read the
result of (depends on) another computation. If adding an edge creates a cycle,
then we stop and report an error instead of getting blocked and deadlocked.

A simple optimization is to skip adding nodes if the computation we're depending on was
already finished, since a computation can't finish if it participates in a dependency cycle.

This algorithm is simple and it works, and Memo used it in the past. However,
the resulting DAG was often large, and so cycle detection was taking ~35% of
incremental zero rebuilds. As a further optimisation, we developed another algorithm
described below.
``

Reply:
Thanks, done!